**Yorum ve anlam** <br>
ABD’deki ailelerin yıllık dinlenme/eğlenme harcamalarını bilmek istediğimizi varsayalım. Aşağıdaki modeli tahmin ettik:

harcamalar=873+0.0012∗yıllık_gelir+0.00002∗yıllık_gelir^2−223.57∗cocuk_sahibi
 
Harcama ABD doları cinsinden eğlenme yıllık harcamasıdır, yıllık_gelir ABD doları cinsinden yıllık gelirdir ve cocuk_sahibi çocuklu aileleri gösteren değişkendir. Tahmin edilen katsayıları yorumlayınız. Yorumlarınızın istatistiksel olarak anlamlı olduğundan emin olmak için hangi ek istatistikler verilmelidir? 

Bias=873 <br>
cocuk_sahibi ile hedef değişken negatif etkidedir. <br>
Yıllık gelir arttıkça yapılabilecek harcamalar artmaktadır. <br>
Daha doğru bilgiler için p değerlerinin de bilinmesi gerekir.

In [8]:
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [1]:
import pandas as pd
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

In [4]:
weather_df.columns

Index(['Formatted Date', 'Summary', 'Precip Type', 'Temperature (C)',
       'Apparent Temperature (C)', 'Humidity', 'Wind Speed (km/h)',
       'Wind Bearing (degrees)', 'Visibility (km)', 'Loud Cover',
       'Pressure (millibars)', 'Daily Summary'],
      dtype='object')

In [5]:
weather_df['TGap (C)']=weather_df['Temperature (C)']-weather_df['Apparent Temperature (C)']

In [9]:
Y = weather_df['TGap (C)']
X = weather_df[['Humidity','Wind Speed (km/h)']]
X = sm.add_constant(X)

In [10]:
sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               TGap (C)   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 27 Jan 2022   Prob (F-statistic):               0.00
Time:                        08:39:02   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Tahmin edilen model: 

$$ TGap (C) = -2.4381 + 3.0292 * Humidity + 0.1193	*Wind Speed (km/h) \qquad(denklem.1)$$ <br>
p-değerleri sıfıra çok yakın olduğundan istatistiksel olarak anlamlıdır.<br>
Humidity 'in modele etkisi Wind Speed'e göre daha fazladır.<br>
Hedef ve açıklayıcı değişken arasında pozitif bir ilişki vardır. <br>
Matematiksel olarak, iki değişkenin etkileşimini, çarpımlarını alarak yakalarız. Şimdi, bir etkileşim değişkeni oluşturalım ve modelimize ekleyelim:

In [11]:
Y = weather_df['TGap (C)']
weather_df['HumidityWind'] = weather_df['Humidity'] * weather_df['Wind Speed (km/h)']
X = weather_df[['Humidity','Wind Speed (km/h)','HumidityWind']]

X = sm.add_constant(X)

sonuclar = sm.OLS(Y, X).fit()
sonuclar.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               TGap (C)   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Thu, 27 Jan 2022   Prob (F-statistic):               0.00
Time:                        08:52:16   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
HumidityWind          0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Tahmin edilen model: 

$$ TGap (C) = 	-0.0839 	-0.1775 * Humidity 	-0.0905	*Wind Speed (km/h) +	0.2971*HumidityWind	 \qquad(denklem.2)$$ <br>
Öncelikle, tüm katsayıların önceki modele göre değişti. <br>
Bunun nedeni, bir değişkeni her eklediğimiz veya çıkarttığımız zaman, model spesifikasyonu değişir. Bu, katsayılar için farklı nokta tahminleri ile sonuçlanır. Aynısı, katsayıların istatistiksel anlamlılığı için de geçerlidir. <br>
Humidity ve Wind Speed modele artık negatif etki yaratmaktadır

**Another Example**

In [63]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

In [64]:
house_prices_df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [65]:
to_drop = ['Id','Alley','PoolQC','Fence','MiscFeature','Electrical']

for k in to_drop:
    house_prices_df.drop([k], axis = 1, inplace = True)


cat_col_data = ['FireplaceQu','GarageType','GarageFinish','MasVnrType','BsmtQual',
           'BsmtCond','BsmtExposure','BsmtFinType1','BsmtFinType2','FireplaceQu',
          'GarageQual','GarageCond']

ncat_col_data = ['LotFrontage','GarageYrBlt','MasVnrArea']
for i in cat_col_data:
    house_prices_df[i] = house_prices_df[i].fillna(house_prices_df[i].mode()[0])
    
for j in ncat_col_data:
    house_prices_df[j] = house_prices_df[j].fillna(house_prices_df[j].mean())

In [66]:
categorical_features=[feature for feature in house_prices_df.columns if house_prices_df[feature].dtype=='O']
for feature in categorical_features:
    labels_ordered=house_prices_df.groupby([feature])['SalePrice'].mean().sort_values().index
    labels_ordered={k:i for i,k in enumerate(labels_ordered,0)}
    house_prices_df[feature]=house_prices_df[feature].map(labels_ordered)


In [43]:
X = house_prices_df.drop(['SalePrice'], axis = 1)
Y = house_prices_df['SalePrice']


In [ ]:
X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

P değerleri incelendiğinde;
MSZoning,LotShape,Exterior2nd,Exterior1st,ExterCond,BsmtCond,YrSold,MiscVal,PoolArea,EnclosedPorch,OpenPorchSF,GarageCond,GarageArea,GarageFinish gibi değerler istatiksel olarak anlamsızdır.

In [56]:
X = house_prices_df.drop(['MSZoning','LotShape','Exterior2nd','Exterior1st','ExterCond','BsmtCond','YrSold','MiscVal','PoolArea','EnclosedPorch','OpenPorchSF','GarageCond','GarageArea','GarageFinish','SalePrice'], axis = 1)
Y = house_prices_df['SalePrice']

In [ ]:
X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

Anlamsız değişkenleri çıkarmama rağmen henüz bir etki yaratmadı.Korele featurelara bakıp modeli ona göre oluşturacağım.

In [59]:
numericData= house_prices_df.select_dtypes(include=[np.number])
_data = numericData.corr()['SalePrice'][:-1] # -1 because the latest row is SalePrice
golden_features_list = _data[abs(_data) > 0.5].sort_values(ascending=False)
print("There is {} strongly correlated values with SalePrice:\n{}".format(len(golden_features_list), golden_features_list))

There is 16 strongly correlated values with SalePrice:
OverallQual    0.791
GrLivArea      0.709
Neighborhood   0.697
ExterQual      0.683
KitchenQual    0.660
BsmtQual       0.650
GarageCars     0.640
GarageArea     0.623
TotalBsmtSF    0.614
1stFlrSF       0.606
FullBath       0.561
GarageFinish   0.537
TotRmsAbvGrd   0.534
YearBuilt      0.523
YearRemodAdd   0.507
Foundation     0.506
Name: SalePrice, dtype: float64


In [69]:
X = house_prices_df[['OverallQual','GrLivArea','Neighborhood','ExterQual']]
Y = house_prices_df['SalePrice']
X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.771
Model:                            OLS   Adj. R-squared:                  0.771
Method:                 Least Squares   F-statistic:                     1227.
Date:                Thu, 27 Jan 2022   Prob (F-statistic):               0.00
Time:                        09:49:08   Log-Likelihood:                -17467.
No. Observations:                1460   AIC:                         3.494e+04
Df Residuals:                    1455   BIC:                         3.497e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -7.306e+04   4802.009    -15.215      0.000   -8.25e+04   -6.36e+04
OverallQual   1.723e+04   1248.713     13.802      0.000    1.48e+04    1.97e+04
GrLivArea       54.7390      2.355     23.244      0.000      50.119      59.358
Neighborhood  3160.5682    217.435     14.536      0.000    2734.049    3587.087
ExterQual     2.011e+04   2634.195      7.632      0.000    1.49e+04    2.53e+04
==============================================================================
Omnibus:                      484.239   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            10348.009
Skew:                           1.017   Prob(JB):                         0.00
Kurtosis:                      15.883   Cond. No.                     7.92e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.92e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Ev fiyatları ile ilişkli özelliklerle model oluşturduğumda R^2 değeri daha kabul edilebilir çıktı ve özellikler ile hedef değişken arasında pozitif bir ilişki mevcut, istatistiksel olarak anlamsız bir değer yok.